db/commitlog/commitlog.cc

```diff
@@ -708,40 +733,38 @@ class db::commitlog::segment : public enable_shared_from_this<segment>, public c
     }
     future<sseg_ptr> close() {
         _closed = true;
-        return sync().then([] (sseg_ptr s) {
-            return s->flush();
-        }).then([](sseg_ptr s) {
-            return s->terminate();
-        }).then([](sseg_ptr s) {
-            s->_segment_manager->totals.wasted_size_on_disk += (s->_size_on_disk - s->file_position());
-            return s;
-        });
+        auto s = co_await sync();
+        co_await flush();
+        co_await terminate();
+        _segment_manager->totals.wasted_size_on_disk += (_size_on_disk - file_position());
+        co_return s;
     }
     future<sseg_ptr> do_flush(uint64_t pos) {
         auto me = shared_from_this();
```

Comment:
Why do we need this?
How can `*this` get destroyed during the coroutine lifetime?

Reply:
If noone holds a reference to it, it can very much be destroyed by someone popping the segment from the segment list. In this particular routine, it will not happen, but symmetry. As stated, this does not aim to refactor behaviour, only coroutinize things.